src/track/beatmap.cpp

```diff
@@ -416,22 +434,14 @@ double BeatMap::getBpm() const {
     if (!isValid()) {
         return -1;
     }
-    return m_dCachedBpm;
-}
-
-double BeatMap::getBpmRange(double startSample, double stopSample) const {
-    QMutexLocker locker(&m_mutex);
-    if (!isValid()) {
-        return -1;
-    }
-    Beat startBeat, stopBeat;
-    startBeat.set_frame_position(
-            static_cast<google::protobuf::int32>(samplesToFrames(startSample)));
-    stopBeat.set_frame_position(static_cast<google::protobuf::int32>(samplesToFrames(stopSample)));
-    return calculateBpm(startBeat, stopBeat);
+    return m_nominalBpm;
 }
 
+// Note: Also called from the engine thread
 double BeatMap::getBpmAroundPosition(double curSample, int n) const {
+    // This function is only usable for one measure or such.
+    DEBUG_ASSERT(n <= 12);
+
     QMutexLocker locker(&m_mutex);
```

Comment:
The mutex could be avoided by copying m_beats and operating on the implicitly-shared, read-only copy instead. This does not introduce any new race conditions since the mutex is unlocked anyway before the result could be returned, i.e. the result could be outdated before it is received by the caller both with and without the mutex.

The same arguments apply to findNthBeat() which involves another unnecessary lock. This member function should be extracted into a free function.

You could then continue with all other (member) functions that work on an immutable BeatList.

Reply:
Getting rid of the mutex here is out of scope of this PR. I am in doubt that this alown helps to get around the locking. 
I think we should make the beat object const, than there is no doubt that the engine can safely read from it without locking.